### ISSUE TITLE
if your server goes down, when it comes back you get messages long before the keys

### ISSUE THREAD

**Issue by turt2live** [matrix-org/matrix-spec#1123]
ref https://github.com/vector-im/element-web/issues/3754

**Comment by MadLittleMods** [comment]
Related to https://github.com/matrix-org/matrix-spec/issues/966 (potential duplicate)

**Comment by richvdh** [comment]
> Related to #966 (potential duplicate)

It's not a duplicate: #966 talks about the ordering of to-device messages relative to one another, while this is about to-device messages relative to room events.

**Comment by dkasak** [comment]
According to https://matrix-org.github.io/synapse/dev-docs/latest/modules/federation_sender.html#a-note-on-failures-and-back-offs, Synapse already mitigates this to some degree by restarting the to-device transmission loop once it receives an inbound request from the remote server.

**Comment by ara4n** [comment]
The core problem here is that todevice messages don't flow transitively, but timeline msgs do. So if server A can't route to server B, messages may all flow transitively A->C->B - but the keys will never get through, no matter what the retry schedule is.

Possible solutions off the top of my head:
 * Forbid transitive backfill for encrypted rooms
 * Let server B request missing todevice traffic from server A via server C. So:
  * server B realises it's receiving current events from server A via server C
  * server B assumes that A can't route to it directly
  * server B asks server C to ask server A for B's todevice msgs.

The latter is more fiddly, but preserves Matrix's somewhat desirable self-healing transitive delivery properties.

**Comment by uhoreg** [comment]
Most of the MLS stuff happens in-room.  The one thing that *can* happen in to-device messages is Welcome messages, but those could also be sent in-room.  We haven't figured out if it would be better to use to-device messages or in-room messages for that.

**Comment by dkasak** [comment]
Presumably Matrix could've gotten away with putting it key data into the room too but for some reason chose not to. Are there any concerns with MLS key data being stored in the DAG forever? E.g. bloat or problems arising out of making permanent something that should be ephemeral?

**Comment by uhoreg** [comment]
Yeah, the main concerns about storing the MLS Welcome messages in the DAG are bloat-related.  It would be irrelevant information to most of the people in the room.

**Comment by uhoreg** [comment]
We may be able to flag events that we got relayed from a different server, so that clients can indicate that there may be a decryption failure.

**Comment by kegsay** [comment]
I think there are two separate issues here:
 - network partitions over federation i.e no amount of waiting will give you the to-device msgs.
 - temporary lag where to-device msgs take ages to arrive.

The latter is quite frequent, and we see this (almost) daily between element.io <--> matrix.org. If we renegotiated room keys less frequently (or had more time to let the room key arrive by not sending it when typing..) this would concretely help I think.

**Comment by uhoreg** [comment]
> (or had more time to let the room key arrive by not sending it when typing..)

I don't understand what you mean by this.  By sending it when typing, that means that we send it earlier, which means it has more time to arrive before the room event is received?

**Comment by BillCarsonFr** [comment]
> Alternatively, does MLS solve this by putting the key data in the timeline rather than todevice msgs?

A document on that subject was created https://docs.google.com/document/d/1f5XMf4qXEzoUBEFDu5IWJXvFAc76D9_dwqt0ReoCRpw/edit#heading=h.a4ncf0eur6o2

**Comment by turt2live** [comment]
MLS key data would most likely be sent over to-device in the linearized matrix models we've been discussing. The guarantees for delivery are a bit higher, though.
